pkg/ebpf/runtime_asset.go

```diff
@@ -0,0 +1,90 @@
+// +build linux_bpf
+
+package ebpf
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/ebpf/compiler"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+)
+
+type CompiledOutput interface {
+	io.Reader
+	io.ReaderAt
+	io.Closer
+}
+
+// RuntimeAsset represents an asset that needs it content integrity checked at runtime
+type RuntimeAsset struct {
+	filename string
+	hash     string
+}
+
+func NewRuntimeAsset(filename, hash string) *RuntimeAsset {
+	return &RuntimeAsset{
+		filename: filename,
+		hash:     hash,
+	}
+}
+
+// Verify reads the asset in the provided directory and verifies the content hash matches what is expected.
+// On success, it returns the full path and content hash of the asset.
+func (a *RuntimeAsset) Verify(dir string) (string, string, error) {
+	p := filepath.Join(dir, "runtime", a.filename)
+	f, err := os.Open(p)
+	if err != nil {
+		return "", "", err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return "", "", fmt.Errorf("error hashing file %s: %w", f.Name(), err)
+	}
+	if fmt.Sprintf("%x", h.Sum(nil)) != a.hash {
+		return "", "", fmt.Errorf("file content hash does not match expected value")
+	}
+	return p, a.hash, nil
+}
+
+// Compile compiles the runtime asset if necessary and returns the resulting file.
+func (a *RuntimeAsset) Compile(config *Config, cflags []string) (CompiledOutput, error) {
+	kv, err := kernel.HostVersion()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get kernel version: %w", err)
+	}
+
+	inputFile, hash, err := a.Verify(config.BPFDir)
+	if err != nil {
+		return nil, fmt.Errorf("error reading input file: %s", err)
+	}
+
+	if err := os.MkdirAll(config.RuntimeCompilerOutputDir, 0755); err != nil {
+		return nil, fmt.Errorf("unable to create compiler output directory %s: %w", config.RuntimeCompilerOutputDir, err)
+	}
+	// filename includes kernel version and input file hash
+	// this ensures we re-compile when either of the input changes
+	baseName := strings.TrimSuffix(a.filename, filepath.Ext(a.filename))
+	outputFile := filepath.Join(config.RuntimeCompilerOutputDir, fmt.Sprintf("%s-%d-%s.o", baseName, kv, hash))
```

Comment:
I wonder if we can take advantage of the source files hash embedded in system-probe to compute a signature (for example with HMAC-256) to bind the compiled object to the source file.

Reply:
Can you elaborate on what you mean here?

Reply:
If I am correct, the current code checks the C files hashes before compiling them to eBPF, but will load any object without any check. Using a signature might be useful to check that a resulting object corresponds to an expected C File.

Reply:
When would the signature be computed and how would it be stored? The object file contents can/will be different on hosts, due to the different kernel versions.

Reply:
It could be computed when compiling the object and stored in the filename as currently done with the hash with something like:
`signature  = HMAC(message=CONTENT_OF_THE_OBJECT, key=C_FILE_HASH)`

This is not a strong security feature, but a proposal to check objects before loading them.

Reply:
One challenge of that approach is currently the expected filename is deterministic from the C file hash. We would need to do some sort of globbing/wildcard approach to find the desired object file.

Since the C files are not embedded in the agent either, I'm not sure a signature check adds much. An attacker could generate an object file with a passing signature. Our currently shipped agent doesn't do any sort of verification of the object files before loading them.

It is worth noting that the system-probe runs as root, and thus these object files would also be owned by root. You would need root access to replace them.